### DISCUSSION TITLE
middleware seems to stop file uploads using API

### DISCUSSION THREAD

**Discussion by redimongo** [vercel/next.js#41277] [Help]
So I created a middleware.ts file because we no longer can put the middleware inside a sub /page/[path] directory, and it seem to stop my upload.js script from running which is hosted in the api folder.

middleware.ts file

```
// middleware.ts
import { NextRequest, NextResponse, userAgent } from 'next/server'

export function middleware(request: NextRequest) {
  const { os } = userAgent(request)


  if(request.nextUrl.pathname === "/appstore")
  {
    if(os.name === "iOS"){
      return NextResponse.redirect('https://apps.apple.com/au/app/drn1/idAPPID')
    }
    if(os.name === "Android"){
      return NextResponse.redirect('https://play.google.com/store/apps/details?id=APP')
    }
    //return NextResponse.rewrite(`/${os}`)
  }
};

export default middleware;

```

Is there something I am missing? As the only way to allow my upload script to work was to remove my middleware file - So the issue is this file is causing the issues.

note my upload file is using Multer and Google storage.

**Comment by PIMI-krl** [comment]
Hi, I've encountered the same issue, my guess is that using middleware results in the request body being "parsed" by the middleware, not in your handler per se, but at some point before your handler actually runs.

I've been able to resolve it by bypassing middleware for the api-endpoint which is responsible for file-upload: e.g. if the enpoint is in `pages/api/upload`, then the following config in `middleware.ts` resolves the issue:

```javascript
export const config = {
  matcher: "/((?!api/upload).*)",
};
```

Hope this helps.

**Reply by redimongo** [reply]
> Hi, I've encountered the same issue, my guess is that using middleware results in the request body being "parsed" by the middleware, not in your handler per se, but at some point before your handler actually runs.
> 
> 
> 
> I've been able to resolve it by bypassing middleware for the api-endpoint which is responsible for file-upload: e.g. if the enpoint is in `pages/api/upload`, then the following config in `middleware.ts` resolves the issue:
> 
> 
> 
> ```javascript
> 
> export const config = {
> 
>   matcher: "/((?!api/upload).*)",
> 
> };
> 
> ```
> 
> 
> 
> Hope this helps.
> 
> 

Yes that works, I forgot to update my question with that answer. 

The issue now will be structuring this middleware in away that works when we need it to work.
